src/node/txreconciliation.cpp

```diff
@@ -0,0 +1,192 @@
+     * - halting of relay of particular transactions due to short ID collisions (DoS)
+     */
+    std::unordered_map<NodeId, uint64_t> m_local_salts GUARDED_BY(m_mutex);
+
+    /**
+     * Keeps track of ongoing reconciliations with a given peer.
+     */
+    std::unordered_map<NodeId, ReconciliationState> m_states GUARDED_BY(m_mutex);
+
+    public:
+
+    std::tuple<bool, bool, uint32_t, uint64_t> PreRegisterPeer(NodeId peer_id, bool peer_inbound)
+    {
+        bool we_initiate_recon, we_respond_recon;
+        // Currently reconciliation roles are defined by the connection direction: only the inbound
+        // peer initiate reconciliations and the outbound peer is supposed to only respond.
+        // If the way we assign reconciliation roles is updated, RECON_VERSION should be bumped.
+        if (peer_inbound) {
+            we_initiate_recon = false;
+            we_respond_recon = true;
+        } else {
+            we_initiate_recon = true;
+            we_respond_recon = false;
+        }
+
+        uint64_t m_local_recon_salt(GetRand(UINT64_MAX));
+        bool added = WITH_LOCK(m_mutex, return m_local_salts.emplace(peer_id, m_local_recon_salt).second);
+        // We do this exactly once per peer (which are unique by NodeId, see GetNewNodeId) so it's
+        // safe to assume we don't have this record yet.
+        assert(added);
+
+        LogPrint(BCLog::NET, "Pre-register peer=%d for reconciling.\n", peer_id);
+        return std::make_tuple(we_initiate_recon, we_respond_recon, RECON_VERSION, m_local_recon_salt);
+    }
+
+    bool RegisterPeer(NodeId peer_id, bool peer_inbound,
+        bool they_may_initiate, bool they_may_respond, uint32_t recon_version, uint64_t remote_salt)
+    {
+        // We do not support reconciliation salt/version updates. We treat an attempt to update
+        // after a successful registration as a protocol violation.
+        LOCK(m_mutex);
+        if (m_states.find(peer_id) != m_states.end()) return false;
+
+        // If the peer supports the version which is lower than our, we downgrade to the version
+        // they support. For now, this only guarantees that nodes with future reconciliation
+        // versions have the choice of reconciling with this current version. However, they also
+        // have the choice to refuse supporting reconciliations if the common version is not
+        // satisfactory (e.g. too low).
+        recon_version = std::min(recon_version, RECON_VERSION);
+        // v1 is the lowest version, so suggesting something below must be a protocol violation.
+        if (recon_version < 1) return false;
+
+        auto local_salt = m_local_salts.find(peer_id);
+
+        // This function should be called only after generating the local salt.
+        if (local_salt == m_local_salts.end()) return false;
+
+        // Must match SuggestReconciling logic.
+        bool we_may_initiate = !peer_inbound, we_may_respond = peer_inbound;
```

Comment:
This is unnecessarily difficult to read. Please don't put multiple declarations/assignments on the same line.